src/Raven.Server/Indexing/VoronBufferedInput.cs

```diff
@@ -0,0 +1,264 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Lucene.Net.Store;
+using Raven.Client.Extensions.Streams;
+using Voron;
+using Voron.Data;
+using Voron.Impl;
+
+namespace Raven.Server.Indexing;
+
+public class VoronBufferedInput : BufferedIndexInput
+{
+    private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+    private const int DefaultBufferSize = 4096;
+    private readonly LuceneVoronDirectory _directory;
+    private readonly string _name;
+    private readonly string _tree;
+    private VoronStream _stream;
+    private bool _isDisposed = false;
+
+    private bool _isOriginal = true;
+
+    /// <summary>Reads an int stored in variable-length format.  Reads between one and
+    /// five bytes.  Smaller values take fewer bytes.  Negative numbers are not
+    /// supported.
+    /// </summary>
+    /// <seealso cref="IndexOutput.WriteVInt(int)">
+    /// </seealso>
+    public override int ReadVInt(IState state)
+    {
+        Optimized:
+        //Perf shortcut
+        if (bufferPosition + 5 < bufferLength)
+        {
+            byte b = buffer[bufferPosition++];
+            int i = b & 0x7F;
+            for (int shift = 7; (b & 0x80) != 0; shift += 7)
+            {
+                b = buffer[bufferPosition++];
+                i |= (b & 0x7F) << shift;
+            }
+
+            return i;
+        }
+
+        //We want to refill only when we're out of cache. Calling this before can lead to data loss. 
+        if (bufferPosition >= bufferLength)
+        {
+            Refill(state);
+
+            // If we got at least 5 elements that means we can go to shortcut.
+            if (bufferPosition + 5 < bufferLength)
+                goto Optimized;
+        }
+
+        //We don't have 5 elements (what is max here) in buffer. To avoid copy pasting of code: go to standard implementation
+
+        return base.ReadVInt(state);
+    }
```

Comment:
```suggestion
    public override int ReadVInt(IState state)
    {
        if (bufferPosition + 5 < bufferLength)
        {
            // handle directly
            byte b = buffer[bufferPosition++];
            int i = b & 0x7F;
            for (int shift = 7; (b & 0x80) != 0; shift += 7)
            {
                b = buffer[bufferPosition++];
                i |= (b & 0x7F) << shift;
            }

            return i;
        }

        return ReadVIntUnlikely(state);
    }

    private int ReadVIntUnlikely(IState state)
    {
        Refill(state);
        if (bufferPosition + 5 < bufferLength)
        {
            return ReadVInt(state);
        }
        // We don't have 5 elements (what is max here) in buffer, just go to standard implementation
        // that would read it one line at a time
        return base.ReadVInt(state);
    }
```

Using this approach, we end us with smaller method for the common case.

Reply:
We might also want to add a check on the `for` loop if the data is wrong, though, no?

Reply:
I don't follow. I see there is a check inside `for` loop

BTW: Suggestion committed.